src/main/java/uk/gov/ida/matchingserviceadapter/services/EidasAssertionService.java

```diff
@@ -0,0 +1,95 @@
+                                 String hubConnectorEntityId,
+                                 String hubEntityId,
+                                 EidasMatchingDatasetUnmarshaller matchingDatasetUnmarshaller) {
+        super(instantValidator, subjectValidator, conditionsValidator, hubSignatureValidator, cycle3DatasetFactory);
+        this.conditionsValidator = conditionsValidator;
+        this.metadataResolverRepository = metadataResolverRepository;
+        this.hubConnectorEntityId = hubConnectorEntityId;
+        this.hubEntityId = hubEntityId;
+        this.matchingDatasetUnmarshaller = matchingDatasetUnmarshaller;
+    }
+
+    @Override
+    void validate(String expectedInResponseTo, List<Assertion> assertions) {
+        Map<Boolean, List<Assertion>> assertionMap = assertions.stream().collect(Collectors.groupingBy(this::isCountryAssertion));
+
+        assertionMap.get(true).forEach(a -> validateCountryAssertion(expectedInResponseTo, a));
+        assertionMap.getOrDefault(false, Collections.emptyList()).forEach(a -> validateCycle3Assertion(a, expectedInResponseTo, hubEntityId));
+    }
+
+    @Override
+    public AssertionData translate(List<Assertion> assertions) {
+        Assertion countryAssertion = assertions.stream().filter(this::isCountryAssertion).findFirst().get();
+        Optional<Assertion> cycle3Assertion = assertions.stream()
+                .filter(a -> !isCountryAssertion(a))
+                .findFirst();
+
+        AuthnStatement authnStatement = countryAssertion.getAuthnStatements().get(0);
+        String levelOfAssurance = authnStatement.getAuthnContext().getAuthnContextClassRef().getAuthnContextClassRef();
+        return new AssertionData(countryAssertion.getIssuer().getValue(),
+                authnContextFactory.mapFromEidasToLoA(levelOfAssurance),
+                getCycle3Data(cycle3Assertion),
+                matchingDatasetUnmarshaller.fromAssertion(countryAssertion));
+    }
+
+    private void validateCountryAssertion(String expectedInResponseTo, Assertion assertion) {
+        metadataResolverRepository.getSignatureTrustEngine(assertion.getIssuer().getValue())
+                .map(MetadataBackedSignatureValidator::withoutCertificateChainValidation)
+                .map(SamlMessageSignatureValidator::new)
+                .map(SamlAssertionsSignatureValidator::new)
+                .orElseThrow(() -> new SamlResponseValidationException("Unable to find metadata resolver for entity Id " + assertion.getIssuer().getValue()))
+                .validate(asList(assertion), IDPSSODescriptor.DEFAULT_ELEMENT_NAME);
+        instantValidator.validate(assertion.getIssueInstant(), "Country Assertion IssueInstant");
+        subjectValidator.validate(assertion.getSubject(), expectedInResponseTo);
+        conditionsValidator.validate(assertion.getConditions(), hubConnectorEntityId);
+    }
+
+    public Boolean isCountryAssertion(Assertion assertion) {
```

Comment:
Might be nicer for this to return one of `enum { COUNTRY, CYCLE_3 }` like in `classifyAssertion` so you don't have to assume if it isn't a country it's a cycle3 in the `groupBy` above (which is probably never going to break, but feels fragile anyway).

Reply:
I came back to this one when working through some of the test. We wanted to get a better error message back when an eidas assertion was signed by a country not in the trust anchor. It's in the rewrite branch.